src/main/java/org/jboss/remoting3/remote/ServerConnectionOpenListener.java

```diff
@@ -258,14 +259,16 @@ public void handleEvent(final ConduitStreamSourceChannel channel) {
                         }
                         final String protocol = optionMap.get(RemotingOptions.SASL_PROTOCOL, RemotingOptions.DEFAULT_SASL_PROTOCOL);
                         final Map<String, String> saslProperties = getSaslProperties(optionMap);
+
                         SaslServer saslServer;
                         try {
-                            saslServer = saslAuthenticationFactory.createMechanism(mechName,
-                                    saslServerFactory -> saslProperties != null ?
-                                            new PropertiesSaslServerFactory(
-                                                    new ProtocolSaslServerFactory(new ServerNameSaslServerFactory(saslServerFactory, serverName), protocol),
-                                                    saslProperties)
-                                            : new ProtocolSaslServerFactory(new ServerNameSaslServerFactory(saslServerFactory, serverName), protocol));
+                            saslServer = saslAuthenticationFactory.createMechanism(mechName, saslServerFactory -> {
+                                saslServerFactory = "EXTERNAL".equals(mechName) ? new SSLSaslServerFactory(saslServerFactory, () -> connection.getSslChannel().getSslSession()) : saslServerFactory;
```

Comment:
We do test this actually, the EXTERNAL mechanism is only in the list of enabled mechanisms if there is a session and if it has a peer.